test/NVQPP/test-0.cpp

```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+// RUN: nvq++ %s -o %basename_t.x --target quantinuum --emulate && ./%basename_t.x | FileCheck %s
+
+#include <cudaq.h>
+#include <iostream>
+
+__qpu__ void bar(cudaq::qubit& q) {
+  x(q);
+}
+
+struct baz {
+  __qpu__ void operator()(cudaq::qubit& q) {
+    x(q);
+  }
+};
+
+struct foo {
+  template <typename CallableKernel>
+  __qpu__ void operator()(CallableKernel &&func) {
+    cudaq::qubit q;
+    func(q);
+    mz(q);
+  }
+};
+
+int main() {
+  // Not supported?:
+  // auto result = cudaq::sample(1000, foo{}, &bar);
```

Comment:
I would think this would work if you change the signature of the foo::operator() to `const CallableKernel&`. As is you'd have to construct the callable in place for the rvalue reference

Reply:
I did try a few variations. The problem is that the argument type won't be a supported argument type, which indicates that everything goes well, until our compiler figure it out that it is not supposed to receive a free function as argument.